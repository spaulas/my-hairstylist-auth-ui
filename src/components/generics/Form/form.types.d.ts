import React from "react";

export type FormProps = {
  children: (
    isFocused: Record<string, boolean>,
    setIsFocused: Dispatch<SetStateAction<{}>>
  ) => React.ReactElement;
  handleSubmit: () => void;
  buttonTitle: string;
  showForgotPassword?: boolean;
  link: { message: string; title: string; url: string };
};
