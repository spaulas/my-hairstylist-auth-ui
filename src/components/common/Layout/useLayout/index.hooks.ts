import { useSelector, useDispatch } from "react-redux";
import { RootReducerState } from "@type/global";
import layoutActions from "@actions/layout/layout.action";

const useLayout = () => {
  const dispatch = useDispatch();

  const { isBackgroundFocused } = useSelector(
    ({ Layout }: RootReducerState) => {
      return { isBackgroundFocused: Layout.isBackgroundFocused };
    }
  );

  const setBackgroundFocus = (focused: boolean) => {
    dispatch(layoutActions.setBackgroundFocus(focused));
  };

  return { isBackgroundFocused, setBackgroundFocus };
};

export default useLayout;
