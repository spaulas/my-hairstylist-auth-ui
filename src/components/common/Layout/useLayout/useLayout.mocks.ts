export const useLayoutMock = jest.fn();

jest.mock("./index.hooks.ts", () => useLayoutMock);
