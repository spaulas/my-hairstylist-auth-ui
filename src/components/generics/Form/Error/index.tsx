import React from "react";
import "./error.styles.scss"

const FormError = ({ showError, message }: any) => {
  return (
    <p className={`error ${showError ? "error__visible" : ""}`}>{message}</p>
  );
};

export default FormError;
