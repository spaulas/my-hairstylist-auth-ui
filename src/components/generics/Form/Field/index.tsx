import React from "react";
import FormError from "../Error";
import "./field.styles.scss";

const FormField = ({
  label,
  errorMessage,
  onChange,
  value,
  type,
  hasError,
  setIsFocused,
  isFocused,
}: any) => {
  const isLabelOnTop = isFocused || value;

  return (
    <>
      <div className="form--item">
        <p
          className={`form--label ${
            isLabelOnTop ? "form--label__focused" : ""
          }`}
        >
          {label}
        </p>
        <input
          type={type}
          name={type}
          id={type}
          className={`form--input ${
            isLabelOnTop ? "form--input__focused" : ""
          }`}
          onFocus={() => setIsFocused(true)}
          onBlur={() => setIsFocused(false)}
          onChange={onChange}
        />
        <FormError showError={hasError} message={errorMessage} />
      </div>
    </>
  );
};

export default FormField;
