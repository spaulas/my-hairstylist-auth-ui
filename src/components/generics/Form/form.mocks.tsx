export const mockSetIsFocused = jest.fn();
export const isFocused = { email: false, password: false };

const Form = ({ children }: any) => {
  return <div>{children(isFocused, mockSetIsFocused)}</div>;
};

jest.mock("./index.tsx", () => ({
  __esModule: true,
  default: Form,
  Form: Form,
}));
