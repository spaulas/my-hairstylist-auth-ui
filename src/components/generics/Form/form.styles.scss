@import "@styles/imports";

.form {
  height: 100%;
  display: flex;
  flex-direction: column;

  &--title {
    margin: 0;
    padding: 104px 0 64px 0;
    text-align: center;
    color: $text--gray__dark;

    &__focused {
      color: $primary--blue;
    }
  }

  &--links {
    &-forgot-password {
      margin-top: 16px;
      display: inline-block;
    }

    &-sign-up {
      @include link;
      margin: auto 0 0 0;
      text-align: center;
    }
  }

  button {
    margin-top: 16px;
  }
}

@media only screen and (max-width: $tablet-landscape__max) {
  .form {
    &--title {
      padding: 82px 0 50px 0;
    }
  }
}

@media only screen and (max-width: $medium-desktop__max) {
  .form {
    &--title {
      padding: 68px 0 42px 0;
    }

    &--links {
      &-forgot-password {
        margin-top: 6px;
      }
    }

    button {
      margin-top: 8px;
    }
  }
}


@media only screen and (max-width: $tablet-portrait__max) {
  .form {
    &--title {
      padding: 60px 0 36px 0;
    }

    &--links {
      &-forgot-password {
        margin-top: 6px;
      }
    }

    button {
      margin-top: 8px;
    }
  }
}
