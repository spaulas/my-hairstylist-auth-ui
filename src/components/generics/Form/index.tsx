import React, { useEffect, useState } from "react";
import "./form.styles.scss";
import Button from "@components/common/Button";
import Link from "@components/common/Link";
import useLayout from "@components/common/Layout/useLayout/index.hooks";

const Form = ({
  children,
  handleSubmit,
  buttonTitle,
  showForgotPassword,
  link: { message, title, url },
}: any) => {
  const { isBackgroundFocused, setBackgroundFocus } = useLayout();

  const [isFocused, setIsFocused] = useState({});

  const isAnyFocused: any = Object.values(isFocused).find((e) => e);

  useEffect(() => {
    if (isAnyFocused !== isBackgroundFocused) {
      setBackgroundFocus(isAnyFocused);
    }
  }, [isFocused]);

  return (
    <div className="form">
      <h1
        className={`form--title ${isAnyFocused ? "form--title__focused" : ""}`}
      >
        My Hairstylist
      </h1>
      {children(isFocused, setIsFocused)}
      <Button onClick={handleSubmit} title={buttonTitle} type="primary" />
      {showForgotPassword ? (
        <Link
          className="form--links-forgot-password"
          title="Forgot your password?"
        />
      ) : null}
      <p className="form--links-sign-up">
        {message} &nbsp;
        <Link url={url} title={title} />
      </p>
    </div>
  );
};

export default Form;
