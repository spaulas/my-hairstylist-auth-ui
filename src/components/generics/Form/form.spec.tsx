import "@testing-library/jest-dom/extend-expect";
import { mockUseState } from "../../../__mocks__/mock-react";
import { mockUseLayout } from "@components/common/Layout/useLayout/useLayout.mocks";
import { fireEvent, render, screen } from "@testing-library/react";
import Form from ".";

describe("Form", () => {
  const childrenText = "Children component";

  const props = {
    children: () => <div>{childrenText}</div>,
    handleSubmit: jest.fn(),
    buttonTitle: "Button title",
    showForgotPassword: false,
    link: { message: "link message", title: "link title", url: "/sign-in" },
  };

  const isFocused = {};
  const mockSetIsFocused = jest.fn();

  const isBackgroundFocused = false;
  const mockSetBackgroundFocus = jest.fn();

  beforeEach(() => {
    mockUseLayout.mockReturnValue({
      isBackgroundFocused,
      setBackgroundFocus: mockSetBackgroundFocus,
    });
    mockUseState.mockReturnValue([isFocused, mockSetIsFocused]);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const { container } = render(<Form {...props} />);
    expect(container).toMatchSnapshot();
  });

  describe("setBackgroundFocus", () => {
    it("should not be called when !isFocused and !isBackgroundFocused", () => {
      render(<Form {...props} />);
      expect(mockSetBackgroundFocus).not.toHaveBeenCalled();
    });

    it("should be called when isFocused and !isBackgroundFocused", () => {
      mockUseState.mockReturnValue([{ email: true }, mockSetIsFocused]);
      render(<Form {...props} />);
      expect(mockSetBackgroundFocus).toHaveBeenCalledWith(true);
    });
  });

  describe("should render correct elements", () => {
    it("title", () => {
      render(<Form {...props} />);
      const title = screen.getByText("My Hairstylist");
      expect(title).toBeVisible();
    });

    it("children", () => {
      render(<Form {...props} />);
      const children = screen.getByText(childrenText);
      expect(children).toBeVisible();
    });

    it("button", () => {
      render(<Form {...props} />);
      const button = screen.getByText(props.buttonTitle);
      expect(button).toBeVisible();
    });

    describe("forgotPassword", () => {
      it("should be visible when showForgotPassword", () => {
        const showForgotPasswordProps = {
          ...props,
          showForgotPassword: true,
        };
        render(<Form {...showForgotPasswordProps} />);
        const forgotPassword = screen.getByText("Forgot your password?");
        expect(forgotPassword).toBeVisible();
      });

      it("should not be visible when !showForgotPassword", () => {
        render(<Form {...props} />);
        const forgotPassword = screen.queryByText("Forgot your password?");
        expect(forgotPassword).toBeFalsy();
      });
    });

    it("message", () => {
      render(<Form {...props} />);
      const message = screen.getByText(props.link.message);
      expect(message).toBeVisible();
    });

    it("link", () => {
      render(<Form {...props} />);
      const linkTitle = screen.getByText(props.link.title);
      expect(linkTitle).toBeVisible();
      expect(linkTitle).toHaveAttribute("href", props.link.url);
    });
  });

  it("should call handleSubmit when button is clicked", () => {
    render(<Form {...props} />);
    const button = screen.getByText(props.buttonTitle);

    fireEvent.click(button);
    expect(props.handleSubmit).toHaveBeenCalled();
  });

  describe("title className", () => {
    it("should have form--title__focused when isAnyFocused", () => {
      render(<Form {...props} />);
      const title = screen.getByText("My Hairstylist");

      expect(title).toHaveClass("form--title");
      expect(title).not.toHaveClass("form--title__focused");
    });

    it("should not have form--title__focused when !isAnyFocused", () => {
      mockUseState.mockReturnValue([{ email: true }, mockSetIsFocused]);
      render(<Form {...props} />);
      const title = screen.getByText("My Hairstylist");

      expect(title).toHaveClass("form--title");
      expect(title).toHaveClass("form--title__focused");
    });
  });
});
