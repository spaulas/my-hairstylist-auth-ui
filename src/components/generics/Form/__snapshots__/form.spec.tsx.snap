// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
<div>
  <div
    class="form"
  >
    <h1
      class="form--title "
    >
      My Hairstylist
    </h1>
    <div>
      Children component
    </div>
    <button
      class="primary"
    >
      Button title
    </button>
    <p
      class="form--links-sign-up"
    >
      link message
       Â 
      <a
        class="link "
        href="/sign-in"
      >
        link title
      </a>
    </p>
  </div>
</div>
`;
